### PR TITLE
add validator weight cap

### PR DESCRIPTION
**Pull request by sampocs**
## Context
Caps the weight percentage that a single validator can have.

**Question for reviewer**: At what number of validators should we start doing this check? I put it at 10 for now, but considering we add all validators in one tx, it actually probably doesn't need to be that high.

## Brief Changelog
* Added `ValidatorWeightCap` param to dictate the max weight percentage
* Added helper function to determine if any validator weights exceed the cap
* Added the check on weights to `AddValidator` and `ChangeValidatorWeights` txs

## Testing
* Commented out [[this](https://github.com/Stride-Labs/stride/blob/main/dockernet/src/register_host.sh#L62-L63)] line from dockernet to prevent any validators from being added at startup
* Started with this json to add validators:
```json
{
	"validators": [
		{
			"name": "val1",
			"address": "cosmosvaloper1tflk30mq5vgqjdly92kkhhq3raev2hnz6eete3",
			"weight": 1
		},
		{
			"name": "val2",
			"address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
			"weight": 1
		},
		{
			"name": "val3",
			"address": "cosmosvaloper1lzhlnpahvznwfv4jmay2tgaha5kmz5qxerarrl",
			"weight": 1
		},
		{
			"name": "val4",
			"address": "cosmosvaloper16k579jk6yt2cwmqx9dz5xvq9fug2tekvlu9qdv",
			"weight": 1
		},
		{
			"name": "val5",
			"address": "cosmosvaloper1g48268mu5vfp4wk7dk89r0wdrakm9p5xk0q50k",
			"weight": 1
		},
		{
			"name": "val6",
			"address": "cosmosvaloper14kn0kk33szpwus9nh8n87fjel8djx0y070ymmj",
			"weight": 1
		},
		{
			"name": "val7",
			"address": "cosmosvaloper10e4vsut6suau8tk9m6dnrm0slgd6npe3jx5xpv",
			"weight": 1
		},
		{
			"name": "val8",
			"address": "cosmosvaloper1vvwtk805lxehwle9l4yudmq6mn0g32px9xtkhc",
			"weight": 1
		},
		{
			"name": "val9",
			"address": "cosmosvaloper124maqmcqv8tquy764ktz7cu0gxnzfw54n3vww8",
			"weight": 1
		},
		{
			"name": "val10",
			"address": "cosmosvaloper140l6y2gp3gxvay6qtn70re7z2s0gn57zfd832j",
			"weight": 1
		}
	]
}
```
* Added validators with the following command and confirmed they were added successfully
```bash
build/strided --home dockernet/state/stride1 tx stakeibc add-validators GAIA validators.json --gas 1000000 --from admin -y
```
* Tried to change a validator's weight
```bash
stridedl tx stakeibc change-validator-weight GAIA cosmosvaloper1tflk30mq5vgqjdly92kkhhq3raev2hnz6eete3 2 --from admin -y
```
* Confirmed it failed
```
 unable to change validator
  weight: validator cosmosvaloper1tflk30mq5vgqjdly92kkhhq3raev2hnz6eete3 exceeds weight
  cap, has 18.181818181818183% of the total weight when the cap is 10%: validator exceeds weight cap
```
* Changed the validators.json file to have one of the weights be 2, then restarted the network and reran the command to add them, and confirmed it errored

### CODE REVIEW

**Review by shellvish — COMMENTED** [review]
This looks great, thank you for doing this Sam! Excellent test coverage as well.

One tricky edge case: if we're in a situation where there are > 30 validators and `ValidatorWeightCap` is already exceeded (e.g. on some of the host zones, we might be in a position where the top 2 validators have > 10% of stake each), I think we wouldn't be able to add any validators or modify weights on the existing validators (because any modification would still fail the `ValidatorWeightCap` check).

Two potential solutions I can think of (but there might be others too!):
1. When we grab `ValidatorWeightCap`, we can wrap it with a helper function that'll return the greater of `ValidatorWeightCap` and the highest validator weight currently
2. When we set weights for a validator, we only check if that specific validator's weight exceeds the `ValidatorWeightCap`

What do you think of these? Also open to any other solutions!

File: proto/stride/stakeibc/params.proto [hunk 1]
```diff
@@ -26,6 +26,7 @@ message Params {
   uint64 default_max_redemption_rate_threshold = 15;
   uint64 ibc_transfer_timeout_nanos = 16;
   uint64 validator_slash_query_threshold = 19;
+  uint64 validator_weight_cap = 20;
```
[the remark this answers is not in the document]
Comment (sampocs) [reply proto/stride/stakeibc/params.proto:29]: yeah I noted it [here](https://github.com/Stride-Labs/stride/pull/1032/files#diff-50de9d0087b725c2a5f237cb915237a29606d28304b2a56803fca3d8c1dbf403R29) where we have our other docs but lmk if you think we should put it some place else!

File: x/stakeibc/keeper/host_zone.go [hunk 2]
```diff
@@ -313,6 +318,42 @@ func (k Keeper) RemoveValidatorFromHostZone(ctx sdk.Context, chainId string, val
...
+		totalWeight += float64(validator.Weight)
+	}
+
+	// If the total validator weights are 0, exit prematurely
+	if totalWeight == 0 {
+		return nil
+	}
+
+	// Check if any validator exceeds the cap
+	for _, address := range utils.StringMapKeys[float64](weightsByValidator) {
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/host_zone.go:345]: this was because of the non-determinism issue with maps in go!

I actually don't think this would re-create the bug cause in the past I think we've only seen it if it could lead to different state transitions.

But considering this throws an error and that error could be different depending on how it loops (imagine a scenario where multiple validators exceed the cap), it felt _slightly_ safe to do the sorting first

File: x/stakeibc/keeper/host_zone.go [hunk 3]
```diff
@@ -274,7 +278,8 @@ func (k Keeper) AddValidatorToHostZone(ctx sdk.Context, chainId string, validato
 
 	k.SetHostZone(ctx, hostZone)
 
-	return nil
+	// Finally, confirm none of the validator's exceed the weight cap
+	return k.CheckValidatorWeightsBelowCap(ctx, hostZone.Validators)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/host_zone.go:282]: It's funny you mentioned this, cause this jumped out to me as well haha.

But I think in general doing just a return is the preferred approach in go (I see it quite often across other code bases) as doing a separate if statement just adds extra code and is essentially the same as doing:
```go
x := someCalculation()
if x != 5 {
   return x
}
return 5
```
(obviously a contrived example haha)

---

And all state is reverted automatically if something fails! That said, I do think it's good practice from a readability perspective to move all checks as upstream as possible, so I'll make the change for that reason!

File: x/stakeibc/keeper/host_zone_test.go [hunk 4]
```diff
@@ -370,3 +370,129 @@ func (s *KeeperTestSuite) TestAddDelegationToValidator() {
...
+
+func (s *KeeperTestSuite) TestCheckValidatorWeightsBelowCap() {
+	testCases := []struct {
+		name       string
+		weightCap  uint64
+		validators []*types.Validator
+		exceedsCap bool
+	}{
+		{
+			name:      "not enough validators",
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/host_zone_test.go:382]: good callout! I'll actually just update this test (I don't think we need a case where weights are 0 cause the check is simply on the length of the array)

File: x/stakeibc/keeper/host_zone.go [hunk 5]
```diff
@@ -313,6 +318,42 @@ func (k Keeper) RemoveValidatorFromHostZone(ctx sdk.Context, chainId string, val
...
 
+// Checks if any validator's portion of the weight is greater than the cap
+func (k Keeper) CheckValidatorWeightsBelowCap(ctx sdk.Context, validators []*types.Validator) error {
+	// If there's only a few validators, don't enforce this yet
+	if len(validators) < MinValidatorsBeforeWeightCapCheck {
+		return nil
+	}
+
+	params := k.GetParams(ctx)
+	validatorWeightCap := float64(params.ValidatorWeightCap)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/host_zone.go:334]: done [[addressed vishal PR comments](https://github.com/Stride-Labs/stride/pull/1032/commits/39149ed73a0a1eadeaf8ff837f017ff261407a14)]

File: x/stakeibc/keeper/host_zone.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/host_zone.go]: done [[addressed vishal PR comments](https://github.com/Stride-Labs/stride/pull/1032/commits/39149ed73a0a1eadeaf8ff837f017ff261407a14)]

**Comment by sampocs** [comment]
> One tricky edge case: if we're in a situation where there are > 30 validators and ValidatorWeightCap is already exceeded (e.g. on some of the host zones, we might be in a position where the top 2 validators have > 10% of stake each), I think we wouldn't be able to add any validators or modify weights on the existing validators (because any modification would still fail the ValidatorWeightCap check).

Wdyt about adding an `update-validator-weight**S**` (plural) tx that takes a json similar to add-validators so we could just update multiple validator weights at once?

File: x/stakeibc/keeper/host_zone.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (shellvish) [reply x/stakeibc/keeper/host_zone.go]: This makes sense as a norm, the example is very helpful! Thanks a lot for making the change!

File: proto/stride/stakeibc/params.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (shellvish) [reply proto/stride/stakeibc/params.proto:29]: Ah that makes sense, let's leave it in there!

File: x/stakeibc/keeper/host_zone.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shellvish) [reply x/stakeibc/keeper/host_zone.go:350]: Yeah that makes sense, let's do it this way!

File: x/stakeibc/keeper/host_zone_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (shellvish) [reply x/stakeibc/keeper/host_zone_test.go:382]: Agreed, that makes sense to me!

File: x/stakeibc/keeper/host_zone.go [hunk 6]
```diff
@@ -313,6 +318,42 @@ func (k Keeper) RemoveValidatorFromHostZone(ctx sdk.Context, chainId string, val
...
+
```
[the remark this answers is not in the document]
Comment (shellvish) [reply x/stakeibc/keeper/host_zone.go:335]: Agreed on putting it in the param validation! imo it's reasonable to exclude, I don't think we'll run into this issue, and it might be slightly better to allow governance to make this param arbitrary

**Comment by shellvish** [comment]
That function makes sense to me, and I think would solve the problem! If we add `update-validator-weights` (plural!) then I think there's no issue. 

After adding that function, I'm all signed off on this PR!

**Comment by sampocs** [comment]
The testing method for the change-validator-weights tx was as follows:

1. Added 10 validators with weight 1
2. Tried to change just one validator weight (using the old tx) - confirmed it failed
3. Tried to change just one validator weight using the json file - confirmed it also failed
4. Tried to change two validator weights using the json file - confirmed it also failed
5. Tried to change all 10 validators to weight 2 using the json file - confirmed it succeeded

---
Example of weight change JSON
```json
{
	"validator_weights": [
		{
			"address": "cosmosvaloper1tflk30mq5vgqjdly92kkhhq3raev2hnz6eete3",
			"weight": 2
		},
		{
			"address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
			"weight": 2
		}
	]
}
```

**Review by shellvish — APPROVED** [review]
The new function looks great, I'm all signed off now! Thank you for doing this!
